lang/syn/src/codegen/program/handlers.rs

```diff
@@ -91,6 +91,16 @@ pub fn generate(program: &Program) -> proc_macro2::TokenStream {
         }
     };
 
+    let non_inlined_event: proc_macro2::TokenStream = {
+        quote! {
+            #[inline(never)]
+            #[cfg(not(feature = "no-cpi-events"))]
+            pub fn __event_dispatch(program_id: &Pubkey, accounts: &[AccountInfo], event_data: &[u8]) -> anchor_lang::Result<()> {
+                Ok(())
```

Comment:
Should we be using the instructions sysvar to check that this call was self-referential, i.e., that the previous instruction was to this program?

Reply:
An example https://github.com/Ellipsis-Labs/phoenix-v1/blob/ee9cc0333fbab6ee20861e51201951b3e9c38dc9/src/lib.rs#L104-L124

Reply:
how about two different macros, `emit_cpi!` and `emit_cpi_signed!` ?

I'll look into the introspection stuff. 

I think devs should have options

Reply:
i would say log authority is better, it is also an account to be provided but much simpler to check.
What would be the advantage of the ix introspection?

Reply:
I don't understand why `log_authority` is even needed here?

Are people writing indexers that parse `ix`s from TransactionStatus Metadata separately from their transaction context? Is that the norm? I don't understand how `log_authority` helps indexers? Sure it prevents other programs from invoking that instruction, but it will still show up in getSignaturesForAddress. 

Can you provide a concrete example of a program that benefits from having `log_authority` guarding the event ix?

Reply:
You need to ask @jarry-xiao

Reply:
It’s not 100% necessary, but it prevents footguns (which I think is one of the key tenets of anchor).

We’ve chatted about this before, and I gave a concrete example of making a CPI from a rogue program into the log instruction of the target program. The parser can defend against this but the logic becomes more error prone.

I think it’s fine if you provide an indexing example of how one might process this along with a test case of how it blocks out erroneous instances of calling the log instruction (both through top level transaction calls and CPIs from rogue programs)

Reply:
Also log_authority means you don’t need introspection, which I think is a win. To my knowledge, you still need to explicitly pass in SysvarInstructions

Reply:
Here’s a super concrete example. Suppose you have a target program T and a rogue program R. R first makes a normal call to T, which will emit a regular event. Then in the same instruction it makes a call to T’s log instruction. You can figure out this is erroneous if you have the tx stack depth, but I think it’s ambiguous in the current interface.

In a vacuum you also know nothing about T’s logging patterns. Maybe it calls emit_cpi! a variable number of times. The point is there’s no way to know (again, until stack depth is in the transaction object which could realistically take months)

However if R’s call to T’s log instruction fails 100% of the time, then there’s nothing to ever worry about

Reply:
Another point for the log authority, if your program allows arbitrary cpi (or only arbitrary self cpi) for any reason (flash loan capability...), then someone can write random logs without the log authority.